### PR TITLE
Upgrade cadl libraries

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix https://github.com/Azure/autorest.csharp/issues/2593
Fix https://github.com/Azure/autorest.csharp/issues/2582

- upgrade cadl compiler and rest version
- update cadl files to comply with the cadl rest
- update emitter to comply with cadl compiler APIs
- support Dictionary Type
- set global apiVersion
- set model Accessibility undefined if not defined in cadl
- update property name 'IsExtendable' to 'IsExtensible' in autorest.csharp

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 1]
```diff
@@ -250,8 +249,25 @@ export function getInputType(
             }
         }
 
-        if (intrinsicName && intrinsicName === "Map") {
-            return getInputTypeForMap(program, m);
+        // if (intrinsicName && intrinsicName === "Map") {
+        //     return getInputTypeForMap(program, m);
+        // }
+        if (m.indexer) {
+            if (isNeverType(m.indexer.key)) {
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:256]: Can we directly use the below directly?

if (!isNeverType(m.indexer.key)){
    do sth. 
}

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 2]
```diff
@@ -14,6 +13,7 @@ import {
     isIntrinsic,
     ModelType,
     ModelTypeProperty,
+    NeverType,
```
Comment (archerzz) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:16]: what does `NeverType` mean?

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (archerzz) [inline test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs:1]: Extensible enum should not have serialization class. See: https://github.com/Azure/autorest.csharp/issues/2528

Also check the test case: https://github.com/Azure/autorest.csharp/blob/79ca039d08177cc16672dace1161c7fd68ec9e9b/test/AutoRest.TestServerLowLevel.Tests/LowLevel/Generation/EnumWritingTests.cs#L11-L15

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the MIT License.
+
+// <auto-generated/>
+
+#nullable disable
+
+namespace CadlFirstTest
+{
+    /// <summary> Extensible enum. </summary>
+    public enum ExtensibleEnum
```
Comment (archerzz) [inline test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.cs:11]: This is a fixed enum. We should rename it.

**Review by archerzz — DISMISSED** [review]
Need to fix the enum name. Others look good to me.

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 2 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:16]: It's cadl Intrinsic type. The never type indicates the values that will never occur.

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs (hunk 3 above)
Comment (chunyu3) [reply test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs:1]: It is extensible enum

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.cs (hunk 4 above)
Comment (chunyu3) [reply test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.cs:11]: It is extensible enum.

File: src/CADL.Extension/Emitter.Csharp/src/emitter.ts [hunk 5]
```diff
@@ -205,11 +189,36 @@ function createModel(program: Program): any {
...
+        if (ids.length > 1) {
+            return ids.slice(0, -2).join("_");
+        }
+    }
+
+    if (operation.interface) {
+        return operation.interface.name;
+    }
+    let namespace = operation.namespace;
+    if (!namespace) {
```
[the remark this answers is not in the document]
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/emitter.ts:208]: after we re-get global namespace or service namespace, the namespace still can be undefined. so need link 214 check at last.

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs (hunk 3 above)
Comment (lirenhe) [reply test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs:1]: @archerzz, as Extensible enum should not have serialization class, why this class get auto generated?

File: test/TestProjects/FirstTest-Cadl/Generated/cadl.json [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+     "Name": "Two",
+     "Value": "2"
+    },
+    {
+     "$id": "9",
+     "Name": "Four",
+     "Value": "4"
+    }
+   ],
+   "IsExtensible": true,
```
Comment (AlexanderSher) [inline test/TestProjects/FirstTest-Cadl/Generated/cadl.json:58]: Hm... in the `InputEnumType`, the property is called `IsExtendable`. @annelo-msft, what do you think should be a correct name for the property?

Comment (AlexanderSher) [reply test/TestProjects/FirstTest-Cadl/Generated/cadl.json:58]: @chunyu3, please rename `IsExtendable` properties into `IsExtensible` in autorest.csharp source

Comment (chunyu3) [reply test/TestProjects/FirstTest-Cadl/Generated/cadl.json:58]: I updated `IsExtendable` to `IsExtensible` in autorest.csharp and fresh the cadl test project.

File: test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs (hunk 3 above)
Comment (chunyu3) [reply test/TestProjects/FirstTest-Cadl/Generated/ExtensibleEnum.Serialization.cs]: resolved

**Review by AlexanderSher — APPROVED** [review]

**Pull request merged**